src/idl/src/directive.c

```diff
@@ -101,150 +109,168 @@ push_source(idl_pstate_t *pstate, const char *inc, const char *abs, bool sys)
   return IDL_RETCODE_OK;
 }
 
-#define START_OF_FILE (1u<<0)
-#define RETURN_TO_FILE (1u<<1)
-#define SYSTEM_FILE (1u<<2)
-#define EXTRA_TOKENS (1u<<3)
-
 static void delete_line(void *ptr)
 {
   struct line *dir = (struct line *)ptr;
   assert(dir);
+  if (dir->path)
+    free(dir->path);
   if (dir->file)
     free(dir->file);
   free(dir);
 }
 
-static idl_retcode_t
-push_line(idl_pstate_t *pstate, struct line *dir)
+static idl_retcode_t push_line(idl_pstate_t *pstate, struct line *dir)
 {
-  idl_retcode_t ret = IDL_RETCODE_OK;
+  idl_retcode_t ret;
 
-  assert(dir);
   if (dir->flags & (START_OF_FILE|RETURN_TO_FILE)) {
-    bool sys = (dir->flags & SYSTEM_FILE) != 0;
-    char *norm = NULL, *abs, *inc;
-    abs = inc = dir->file;
-    /* convert to normalized file name */
-    if (!idl_isabsolute(abs)) {
-      /* include paths are relative to the current file. so, strip file name,
-         postfix with "/relative/path/to/file" and normalize */
-      const char *cwd = pstate->scanner.position.source->path->name;
-      const char *sep = cwd;
-      assert(idl_isabsolute(cwd));
-      for (size_t i=0; cwd[i]; i++) {
-        if (idl_isseparator(cwd[i]))
-          sep = cwd + i;
+    char *norm = NULL;
+    const idl_source_t *src = pstate->scanner.position.source;
+
+    if (!idl_isabsolute(dir->path)) {
+      const char *cwd = src->path->name;
+
+      if (cwd && strcmp(cwd, "<builtin>") != 0) {
+        char *abs = NULL;
+        size_t len = 0, sep = 0;
+        assert(idl_isabsolute(cwd));
+        for (size_t pos=0; cwd[pos]; pos++) {
+          if (!idl_isseparator(cwd[pos]))
+            sep = 0;
+          else if (!sep)
+            len = sep = pos;
+        }
+        if (idl_asprintf(&abs, "%.*s", len, cwd, dir->path) < 0)
+          return IDL_RETCODE_NO_MEMORY;
+        free(dir->path);
+        dir->path = abs;
       }
-      if (idl_asprintf(&abs, "%.*s/%s", (sep-cwd), cwd, inc) < 0)
-        return IDL_RETCODE_NO_MEMORY;
     }
-    idl_normalize_path(abs, &norm);
-    if (abs != dir->file)
-      free(abs);
-    if (!norm)
-      return IDL_RETCODE_NO_MEMORY;
 
-    if (dir->flags & START_OF_FILE) {
-      ret = push_source(pstate, inc, norm, sys);
-    } else {
-      assert(pstate->scanner.position.source);
-      const idl_source_t *src = pstate->scanner.position.source;
-      while (src) {
-        if (strcmp(src->path->name, norm) == 0)
+    if ((ret = idl_normalize_path(dir->path, &norm)) < 0)
+      return ret;
+    free(dir->path);
+    dir->path = norm;
+
+    if (dir->flags & RETURN_TO_FILE) {
+      for (; src; src = src->parent)
+        if (src->path->name && strcmp(src->path->name, dir->path) == 0)
           break;
-        src = src->parent;
-      }
+        /* search for source */ ;
```

Comment:
Something got mangled here (wrongly indented comment with a meaningless `;` — one would think it existed at some point as an empty loop body)